### ISSUE TITLE
R crashes if I fit a model, modify the model and fit again twice

### ISSUE THREAD

**Issue by martinmodrak** [stan-dev/rstan#399]
#### Summary:
R frequently crashes during model development. Note that this is related to issue #396, but while #396 is resolved when building the package from latest source (as of 2017-03-10), the problem reported here persists. 

#### Description:
If I fit a model, modify the model file and then fit again twice, R crashes. The crash is contingent on calling ```extract``` for the fits.

#### Reproducible Steps:
Here is an [SSCCE](http://sscce.org/) :
```
library(rstan)
rstan_options(auto_write = TRUE)

simple_model = "parameters {
  real a;
}
model {
  a ~ normal(0,1); 
}
"

fileConn<-file("test.stan")
writeLines(simple_model, fileConn)
close(fileConn)

a = extract(stan("test.stan"))

fileConn<-file("test.stan")
writeLines(c(simple_model," "), fileConn) #Adding a space at the end of file to force recompilation
close(fileConn)

a = extract(stan("test.stan"))
a = extract(stan("test.stan")) #This line crashes.

```

I have tested and no step can be ommitted - in particular, it is necessary to call ```extract``` on the fits and it is necessary to fit the model twice after a recompilation - first fit after a recompilation will not trigger the bug, neither will multiple fits without recompiling.

#### Expected Output:
Do not crash.

#### RStan Version:
Both 2.14.1 and the current development version (installed via ```devtools::install_github("stan-dev/rstan", ref = "develop", subdir = "rstan/rstan")``` on 2017-03-10)

#### R Version:
R version 3.3.2 (2016-10-31)

#### Operating System:
Windows 10

**Comment by bgoodri** [comment]
I believe I have fixed this. Please install the develop branch from GitHub and try on your machine.

In general, things are safer if you assign the output of `stan()` or `sampling()` to an object rather than wrapping it in something like `extract()` and letting the garbage collector deal with the stanfit object.

**Comment by martinmodrak** [comment]
I can verify that the crash is fixed, thanks!

What I however find strange is the need to recompile the model for every execution. Why is it that if I don't modify the file, recompilation is not necessary for any number of executions, but if I modify the file, it will recompile on every subsequent execution (modified or not)?

I tried deleting the .rds file and it does not help, the only way to prevent recompilation seems to be to restart the R session - do you have any ideas on that?

Also to note - the bug occured also in a normal development cycle where I stored the fit in an object and inspected it separately wtih extract. (although I am overriding the variable storing the fit with new values)

**Comment by bgoodri** [comment]
It has to do with the fact that the stanfit object is not assigned to a symbol (and thus gets deleted when it goes out of scope). During that process, the DSO gets loaded into R's memory but in subsequent calls to `sampling` or `stan` there is no stanfit object that we can dig into to get the stanmodel object to reuse. The crash happens because R gets two copies of the DSO or tries to access the outdated one or something. Anyway, it should be fixed now, although there might be excessive recompilation. You can probably avoid that by assigning the result of `stan` to an object before calling `extract` on it and / or passing the stanfit object to the `fit` argument of `stan` in subsequent calls.
